server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotShutdownIT.java

```diff
@@ -0,0 +1,294 @@
+            final var entriesForRepo = SnapshotsInProgress.get(state).forRepo(repoName);
+            if (entriesForRepo.isEmpty()) {
+                return true;
+            }
+            assertThat(entriesForRepo, hasSize(1));
+            final var shardSnapshotStatuses = entriesForRepo.iterator().next().shards().values();
+            assertThat(shardSnapshotStatuses, hasSize(1));
+            assertThat(
+                shardSnapshotStatuses.iterator().next().state(),
+                oneOf(SnapshotsInProgress.ShardState.INIT, SnapshotsInProgress.ShardState.SUCCESS)
+            );
+            return false;
+        });
+
+        PlainActionFuture.<Void, RuntimeException>get(
+            fut -> putShutdownMetadata(
+                clusterService,
+                SingleNodeShutdownMetadata.builder()
+                    .setType(SingleNodeShutdownMetadata.Type.RESTART)
+                    .setStartedAtMillis(clusterService.threadPool().absoluteTimeInMillis())
+                    .setReason("test"),
+                originalNode,
+                fut
+            ),
+            10,
+            TimeUnit.SECONDS
+        );
+        assertFalse(snapshotCompletesWithoutPausingListener.isDone());
+        unblockAllDataNodes(repoName); // lets the shard snapshot continue so the snapshot can succeed
+        assertEquals(SnapshotState.SUCCESS, snapshotFuture.get(10, TimeUnit.SECONDS).getSnapshotInfo().state());
+        safeAwait(snapshotCompletesWithoutPausingListener);
+        clearShutdownMetadata(clusterService);
+    }
+
+    public void testRemoveNodeDuringSnapshot() throws Exception {
+        internalCluster().ensureAtLeastNumDataNodes(1);
+        final var originalNode = internalCluster().startDataOnlyNode();
+        final var indexName = randomIdentifier();
+        createIndexWithContent(
+            indexName,
+            indexSettings(1, 0).put(IndexMetadata.INDEX_ROUTING_REQUIRE_GROUP_PREFIX + "._name", originalNode).build()
+        );
+
+        final var repoName = randomIdentifier();
+        createRepository(repoName, "mock");
+
+        final var clusterService = internalCluster().getCurrentMasterNodeInstance(ClusterService.class);
+        final var snapshotFuture = startFullSnapshotBlockedOnDataNode(randomIdentifier(), repoName, originalNode);
+        final var snapshotPausedListener = createSnapshotPausedListener(clusterService, repoName);
+
+        updateIndexSettings(Settings.builder().putNull(IndexMetadata.INDEX_ROUTING_REQUIRE_GROUP_PREFIX + "._name"), indexName);
+        putShutdownForRemovalMetadata(originalNode, clusterService);
+        unblockAllDataNodes(repoName); // lets the shard snapshot abort, which frees up the shard so it can move
+        safeAwait(snapshotPausedListener);
+
+        // snapshot completes when the node vacates even though it hasn't been removed yet
+        assertEquals(SnapshotState.SUCCESS, snapshotFuture.get(10, TimeUnit.SECONDS).getSnapshotInfo().state());
```

Comment:
I think this only completes because there is only one shard. AFAICS, if the snapshot had a shard on an unrelated node, we'd not allow the shard on the shutting down node to move in `SnapshotInProgressAllocationDecider`.

I wonder if we should generally allow moving shards in shard snapshot state `WAITING` in the decider. On the surface seems ok to me, though we could delay the snapshot indefinitely if a shard keeps moving around for unrelated reasons? Alternatively we can for a WAITING shard snapshot check that the current node has a shutdown indication, which I'd probably prefer for now to make it a targetted fix. We could also invent a new state, but seems more hazzle than beneficial.

Reply:
++ added a TODO in c7b312a9bf6b5317671a2893a1dea1edfaf00b87

Reply:
> if the snapshot had a shard on an unrelated node, we'd not allow the shard on the shutting down node to move in SnapshotInProgressAllocationDecider.

Is this because the following line would other include the snapshot entry, find out the "WAITING" shard is incomplete and return `THROTTLE` for allocation?

https://github.com/elastic/elasticsearch/blob/6c8e585f1414b7abd524f02f6a6b45ad19d7a4a7/server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SnapshotInProgressAllocationDecider.java#L70

Preventing a shard from moving on one node because the index has another shard on a different node that is doing work seems to be rather a strong constraint. Stronger than I initially thought about this issue, i.e. I thought a shard cannot be moved if _itself_ is doing snapshoting work. Do we need to relax this constraint in this PR? It seems that the new feature won't be more generally helpful with this constraint.

Also, for my education, how is the `THROTTLE` decision different from `NO` in this case? i.e. why is it not just `NO`?

Reply:
It's `THROTTLE` because the shard snapshot is only a temporary thing. That way the balancer will pretend that it actually moved this shard.